File: utils/odFilter.py
'''
    input: sc.textFile('')
    output:"OD,starttime,endtime"
'''
def ssplit(x):
    L = x.split(',')
    return (L[1],L[4]+','+L[1]+','+L[3]+','+L[6])

def ODRuler(x,y):
    L1 = x.split(',')
    L2 = y.split(',')
    if not(L1[2] == '21' and L2[2] == '22'):
        return None
    if int(L1[0][11:13]) - int(L2[0][11:13])> 3:
        return None
    else :
        return L1[1]+','+L1[0]+','+L1[3]+','+L2[0]+','+L2[3]

def MakeOD(para):
    i = 0
    res = []
    dd = para[1].data
    dd.sort()
    while(i < len(dd)-1):
        od = ODRuler(dd[i],dd[i+1])
        if od is not None:
            res.append(od)
            i = i+2
        else :
            i = i+1
    return res

def ODFilter(data):
    data = data.map(ssplit)
    data = data.groupByKey()
    data = data.flatMap(MakeOD)
    return data

File: scala/src/main/scala/com/sibat/utils/ODFilter.scala
package com.sibat.utils

import java.text.SimpleDateFormat
import org.apache.spark.rdd.RDD

object ODFilter {

  private def ssplit(x:String) = {
    val L = x.split(",")
    (L(0),List(L(1),L(0),L(3),L(2)).mkString(","))
  }

  private def delTime(t1:String,t2:String) = {
    val sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    (sdf.parse(t2).getTime - sdf.parse(t1).getTime) / 1000
  }

  private def ODRuler(x:String,y:String) = {
    val L1 = x.split(",")
    val L2 = y.split(",")
    val difftime = delTime(L1(0),L2(0))
    if(
      ((L1(2) == "21") && (L2(2) == "22")) &&
      ( difftime < 10800) && (L1(3) != L2(3))
    ) List(L1(1),L1(0),L1(3),L2(0),L2(3),(difftime.toString)).mkString(",")
    else None
  }

  private def MakeOD(x:(String,Iterable[String])) = {
    val arr = x._2.toArray.sortWith((x,y) => x < y)
    for{
       i <- 0 until arr.size - 1;
       od = ODRuler(arr(i),arr(i+1))
     } yield od
  }

  def getOD(data:RDD[String]) = {
    data.map(ssplit)
        .groupByKey()
        .flatMap(MakeOD)
        .filter(x => x != None)
  }

}
